lib/synapse/service_watcher/multi/resolver/s3_toggle.rb

```diff
@@ -0,0 +1,229 @@
+      watcher_weights = watcher_weights.select do |watcher, _|
+        exists = @watchers.has_key?(watcher)
+        unless exists
+          log.warn "synapse: s3 toggle resolver: read invalid watcher name #{watcher}"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.unknown_watchers')
+        end
+
+        exists
+      end
+
+      watcher_name = if watcher_weights.length > 0
+        # Pick a watcher randomly by weight.
+        total_weight = watcher_weights.values.inject(:+)
+        pick = rand(total_weight)
+        chosen_watcher = nil
+
+        watcher_weights.each do |key, value|
+          if pick < value
+            chosen_watcher = key
+            break
+          else
+            pick -= value
+          end
+        end
+
+        if chosen_watcher.nil?
+          log.warn "synapse: s3 toggle resolver: failed to pick a watcher"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:fail', 'reason:no_choice'])
+
+          nil
+        else
+          log.info "synapse: s3 toggle resolver: chose watcher #{chosen_watcher}"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:success', "watcher:#{chosen_watcher}"])
+          chosen_watcher
+        end
+      else
+        log.warn "synapse: s3 toggle resolver: no watchers read, failed to pick a watcher"
+        statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:fail', 'reason:watchers_missing'])
+
+        nil
+      end
+
+      return watcher_name
+    end
+
+    def read_s3_file
+      s3 = AWS::S3::Client.new
+
+      data =
+        begin
+          resp = s3.get_object(bucket: @s3_bucket, key: @s3_path)
+          parsed = YAML.load(resp.body.read)
+
+          log.info "synapse: s3 toggle resolver: read s3 file: #{parsed}"
+          parsed
+        rescue Psych::SyntaxError => e
+          log.warn "synapse: s3 toggle resolver: failed to parse s3 file: #{e}"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.fetch_failure', ["reason:parse_error"])
+          nil
+        rescue AWS::Errors::Base => e
+          log.warn "synapse: s3 toggle resolver: failed to fetch s3 file: #{e}"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.fetch_failure', ["reason:s3_error"])
+          nil
+        end
+
+      if validate_s3_file_schema(data)
+        return data
+      else
+        log.warn "synapse: s3 toggle resolver: s3 file has invalid schema"
+        statsd_increment('synapse.watcher.multi.resolver.s3_toggle.fetch_failure', ["reason:invalid_schema"])
+
+        return nil
+      end
+    end
+
+    # expected schema is {'watcher_name' => Integer, ...}
+    def validate_s3_file_schema(contents)
+      return false if (
+          contents.nil? ||
+          !contents.is_a?(Hash)
+        )
+
+      contents.each do |key, value|
+        return false unless key.is_a?(String) && value.is_a?(Integer)
```

Comment:
shall we allow floating type? not a big deal but just want to confirm

Reply:
I don't think we should, it makes the mental math harder 😛 It also doesn't add much value other than flexibility, but I don't think that flexibility is worthwhile